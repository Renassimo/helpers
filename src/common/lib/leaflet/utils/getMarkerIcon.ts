import { DivIcon } from 'leaflet';
import invert from 'invert-color';

const getMarkerIcon = (color = '#ffffff', isNew = false) =>
  new DivIcon({
    html: `<svg xmlns="http://www.w3.org/2000/svg" data-name="Map Pin" viewBox="0 0 128 160" x="0px" y="0px">
	<title>Map Pin</title>
	<path opacity="${isNew ? 0.1 : 1}" fill="${invert(
      color,
      true
    )}" d="M 64 2 C 41.91 2 24 19.08 24 40.15 S 64 126 64 126 s 40 -64.77 40 -85.85 S 86.09 2 64 2 Z Z"/>
	<path opacity="${
    isNew ? 0.55 : 1
  }" fill="${color}" d="M 64.04 3.62 C 42.8336 3.62 25.64 20.0168 25.64 40.244 S 64.04 122.66 64.04 122.66 s 38.4 -62.1792 38.4 -82.416 S 85.2464 3.62 64.04 3.62 Z Z"/>
</svg>`,
    className: 'marker',
    iconSize: [48, 60],
    iconAnchor: [24, 46.5],
  });

export default getMarkerIcon;
