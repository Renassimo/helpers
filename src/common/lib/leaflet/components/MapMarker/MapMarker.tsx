import { Marker, Popup, Tooltip } from 'react-leaflet';
import { ReactNode } from 'react';

import getMarkerIcon from '../../utils/getMarkerIcon';

const MapMarker = ({
  color,
  position,
  isNew = false,
  isMarked = false,
  children,
  title,
}: {
  color: string;
  position: [number, number];
  isNew?: boolean;
  isMarked?: boolean;
  children?: ReactNode;
  title?: string;
}) => {
  return (
    <>
      <Marker position={position} icon={getMarkerIcon(color, isNew, isMarked)}>
        {title && (
          <Tooltip direction="top" offset={[0, -45]}>
            {title}
          </Tooltip>
        )}
        {children && <Popup>{children}</Popup>}
      </Marker>
    </>
  );
};

export default MapMarker;
