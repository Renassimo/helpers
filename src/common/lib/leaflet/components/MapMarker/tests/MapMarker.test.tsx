import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import MapMarker from '../MapMarker';
import getMarkerIcon from '../../../utils/getMarkerIcon';

import { Marker, Popup } from 'react-leaflet';

jest.mock('react-leaflet');
jest.mock('../../../utils/getMarkerIcon');

describe('MapMarker snapshot', () => {
  const MokcedMarker = jest.fn(({ position, children, icon }) => (
    <div>
      MockedMarker - {position[0]} - {position[1]} - {icon} - {children}
    </div>
  ));
  const MockedPopup = jest.fn(({ children }) => (
    <div>MockedPopup - {children}</div>
  ));
  const mockedIcon = 'mocked-icon';
  const mockedGetMarkerIcon = jest.fn(() => mockedIcon);

  beforeEach(() => {
    (Marker as unknown as jest.Mock).mockImplementation(MokcedMarker);
    (Popup as unknown as jest.Mock).mockImplementation(MockedPopup);
    (getMarkerIcon as unknown as jest.Mock).mockImplementation(
      mockedGetMarkerIcon
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders successfully', () => {
    const mockedColor = '#aaaaaa';
    const mockedPosition = [0.1, 10.1] as [number, number];
    // Arange
    // Act
    const { baseElement } = renderWithTheme(
      <MapMarker color={mockedColor} position={mockedPosition} />
    );
    // Assert
    expect(baseElement).toMatchSnapshot();
    expect(mockedGetMarkerIcon).toHaveBeenCalledWith(mockedColor, false);
  });

  describe('when children passed', () => {
    test('renders successfully', () => {
      const mockedColor = '#aaaaaa';
      const mockedPosition = [0.1, 10.1] as [number, number];
      // Arange
      // Act
      const { baseElement } = renderWithTheme(
        <MapMarker color={mockedColor} position={mockedPosition} isNew>
          Popup Content
        </MapMarker>
      );
      // Assert
      expect(baseElement).toMatchSnapshot();
      expect(mockedGetMarkerIcon).toHaveBeenCalledWith(mockedColor, true);
    });
  });
});
