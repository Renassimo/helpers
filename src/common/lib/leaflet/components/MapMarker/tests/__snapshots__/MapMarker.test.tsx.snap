// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapMarker snapshot renders successfully 1`] = `
<body>
  <div>
    <div>
      MockedMarker - 
      0.1
       - 
      10.1
       - 
      mocked-icon
       - 
      <div>
        MockedTooltip - 
        top
         - 
        0
         - 
        -45
      </div>
    </div>
  </div>
</body>
`;

exports[`MapMarker snapshot when children passed renders successfully 1`] = `
<body>
  <div>
    <div>
      MockedMarker - 
      0.1
       - 
      10.1
       - 
      mocked-icon
       - 
      <div>
        MockedPopup - 
        Popup Content
      </div>
    </div>
  </div>
</body>
`;
